### PR TITLE
node-network-attach-error-message

### PR DESCRIPTION
**Pull request by skanthed**

### CODE REVIEW

**Comment by skanthed** [comment]
Sure, In the next commit I will fix it.

File: esiclient/v1/node_network.py [hunk 1]
```diff
@@ -214,27 +214,39 @@ def take_action(self, parsed_args):
...
+                print("Attaching network {1} to node {0}{2}".format(
+                    node.name, network.name, mac_string))
+                port_name = utils.get_port_name(network.name, prefix=node.name)
+                port = utils.get_or_create_port(port_name, network,
+                                                neutron_client)
+                ironic_client.node.vif_attach(node_uuid, port.id, **vif_info)
+                port = neutron_client.get_port(port.id)
+            else:
+                print("The specified network does not exist."
+                      "Please provide a valid network name.")
```
Comment (tzumainn) [inline esiclient/v1/node_network.py:227]: Instead of printing an error message, you'll want to raise `exceptions.CommandError` (you can find examples of this if you grep in the repository). You should also check if the network exists right away - up in line 177.

File: esiclient/v1/node_network.py [hunk 2]
```diff
@@ -214,27 +214,39 @@ def take_action(self, parsed_args):
...
         elif parsed_args.trunk:
             print("Attaching trunk {1} to node {0}{2}".format(
                 node.name, trunk.name, mac_string))
             port = neutron_client.get_port(trunk.port_id)
             ironic_client.node.vif_attach(node_uuid, port.id, **vif_info)
 
-        network_names, port_names, fixed_ips \
-            = utils.get_full_network_info_from_port(
-                port, neutron_client)
+        if port is not None:
```
Comment (tzumainn) [inline esiclient/v1/node_network.py:234]: This `if` won't be needed if you check for the network right away.

File: esiclient/v1/node_network.py [hunk 3]
```diff
@@ -214,27 +214,39 @@ def take_action(self, parsed_args):
...
-            [node.name, port.mac_address,
-             "\n".join(port_names),
-             "\n".join(network_names),
-             "\n".join(fixed_ips)]
+            return ["Node", "MAC Address", "Port", "Network", "Fixed IP"], \
+                [node.name, port.mac_address,
+                 "\n".join(port_names),
+                 "\n".join(network_names),
+                 "\n".join(fixed_ips)]
+        else:
```
Comment (tzumainn) [inline esiclient/v1/node_network.py:244]: We don't really want to return a blank result.

**Review by tzumainn — CHANGES_REQUESTED** [review]
Good start! Comments inline.

File: test-requirements.txt [hunk 4]
```diff
@@ -8,6 +8,7 @@ docutils>=0.11 # OSI-Approved Open Source, Public Domain
 fixtures>=3.0.0 # Apache-2.0/BSD
 metalsmith
 mock>=3.0.0 # BSD
+munch>=3.0.0 # MIT
```
Comment (tzumainn) [inline test-requirements.txt:11]: Why is this needed?

**Review by tzumainn — CHANGES_REQUESTED** [review]
Small inline comment; the bigger item is you'll want to create a unit test covering this change. Let me know if you'd like some guidance regarding how/where to do that!

File: test-requirements.txt (hunk 4 above)
Comment (skanthed) [reply test-requirements.txt:11]: Was getting the error - ModuleNotFoundError: No module named 'munch' .
The specific failure occurred while trying to import the test module esiclient.tests.unit.v1.mdc.test_mdc_node_baremetal.

It seems that the test module test_mdc_node_baremetal.py has a dependency on the munch module, but it is not installed in our test environment.
In earlier PR's, they had different python versions and packages, the packages didn't depend on the munch

Comment (tzumainn) [reply test-requirements.txt:11]: Ah, I see - good catch!

File: esiclient/v1/node_network.py [hunk 5]
```diff
@@ -175,6 +175,9 @@ def take_action(self, parsed_args):
         if parsed_args.network:
             network = neutron_client.find_network(parsed_args.network)
             port = None
+            if network is None:
+                raise exceptions.CommandError(
+                    "ERROR: This is not a network name")
```
Comment (tzumainn) [inline esiclient/v1/node_network.py:180]: I'd say something like 'Unknown network' instead, since I think the user can also pass in a network UUID.

File: esiclient/tests/unit/v1/test_node_network.py [hunk 6]
```diff
@@ -469,6 +473,24 @@ def test_take_action_no_port_or_network_or_trunk_exception(self):
...
+            return_value = self.node
+        self.app.client_manager.baremetal.port.list.\
+            return_value = [self.port1]
+
+        arglist = ['node1', '--network', 'test_network_invalid']
+        verifylist = []
+
+        parsed_args = self.check_parser(self.cmd, arglist, verifylist)
+#       results = self.cmd.take_action(parsed_args)
+#       print("Invalid network exception results {}".format(results))
```
Comment (tzumainn) [inline esiclient/tests/unit/v1/test_node_network.py:487]: Make sure to remove these commented out lines!

**Review by tzumainn — CHANGES_REQUESTED** [review]

File: esiclient/v1/node_network.py (hunk 5 above)
Comment (skanthed) [reply esiclient/v1/node_network.py:180]: I would say something like - "ERROR: This is not a network name or UUID" because then it aligns with the error message we are passing for the wrong port name or trunk name below.

File: esiclient/tests/unit/v1/test_node_network.py (hunk 6 above)
Comment (skanthed) [reply esiclient/tests/unit/v1/test_node_network.py:487]: Sure! They are just for my test purpose.

**Comment by skanthed** [comment]
@tzumainn can you look into the test case of this one? I tried multiple things but still got errors.

File: esiclient/v1/node_network.py (hunk 5 above)
Comment (tzumainn) [reply esiclient/v1/node_network.py:180]: To be honest, I'd change it because whatever the user passes in *is* a network name; it's just not necessarily one that's in the system.

File: esiclient/tests/unit/v1/test_node_network.py [hunk 7]
```diff
@@ -469,6 +473,24 @@ def test_take_action_no_port_or_network_or_trunk_exception(self):
             'ERROR: You must specify either network, port, or trunk',
             self.cmd.take_action, parsed_args)
 
+    def test_take_action_invalid_network_exception(self, mock_gfnifp):
```
Comment (tzumainn) [inline esiclient/tests/unit/v1/test_node_network.py:476]: You're missing the decorator here - actually, you probably don't need mock_gfnifp if you're not using it.

Comment (skanthed) [reply esiclient/tests/unit/v1/test_node_network.py:476]: yes I removed it, I don't think I need to use it here. It was here by mistake because I was trying something when it didn't work.

File: esiclient/tests/unit/v1/test_node_network.py [hunk 8]
```diff
@@ -469,6 +473,22 @@ def test_take_action_no_port_or_network_or_trunk_exception(self):
             'ERROR: You must specify either network, port, or trunk',
             self.cmd.take_action, parsed_args)
 
+    def test_take_action_invalid_network_exception(self):
+        self.app.client_manager.baremetal.node.get.\
+            return_value = self.node
+        self.app.client_manager.baremetal.port.list.\
+            return_value = [self.port1]
+
```
Comment (tzumainn) [inline esiclient/tests/unit/v1/test_node_network.py:481]: Oh, you want to mock the call to `find_network` as well to have it explicitly return `None`.

File: esiclient/tests/unit/v1/test_node_network.py [hunk 9]
```diff
@@ -469,6 +473,33 @@ def test_take_action_no_port_or_network_or_trunk_exception(self):
...
+    #             return_value=(["test_network_invalid"], ["node1-port"],
+    #                           ["2.2.2.2"]),
+    #             autospec=True)
+    def test_take_action_invalid_network_exception(self):
+        self.app.client_manager.baremetal.node.get.\
+            return_value = self.node
+        self.app.client_manager.baremetal.port.list.\
+            return_value = [self.port1]
+        self.app.client_manager.network.find_network.\
+            return_value = self.network
```
Comment (tzumainn) [inline esiclient/tests/unit/v1/test_node_network.py:486]: So what's happening here is that even though you are passing in 'test_network_invalid' - a network that doesn't exist - you are also telling `find_network` to always return a network. Because of that, this test will always think that the network exists. What you want to do is set this return_value to None instead.

File: esiclient/tests/unit/v1/test_node_network.py [hunk 10]
```diff
@@ -282,6 +282,10 @@ def setUp(self):
             "id": "network_uuid",
             "name": "test_network"
         })
+        self.invalid_network = utils.create_mock_object({
+            "id": "network_uuid_invalid",
+            "name": "test_network_invalid"
+        })
```
Comment (tzumainn) [inline esiclient/tests/unit/v1/test_node_network.py:288]: I don't think this is used anywhere, so it's safe to remove.

File: esiclient/v1/node_network.py [hunk 11]
```diff
@@ -174,7 +174,11 @@ def take_action(self, parsed_args):
 
         if parsed_args.network:
             network = neutron_client.find_network(parsed_args.network)
+            print("Network {} and length {}".format(network, type(network)))
```
Comment (tzumainn) [inline esiclient/v1/node_network.py:177]: This should be removed.

**Review by tzumainn — CHANGES_REQUESTED** [review]
Close! Just a few debugging items to remove.

**Comment by tzumainn** [comment]
Looks good - thanks!
